Generic protocols

tommarek opened the pull request:
a rather big change, please review and comment on this if you have any questions or would like something done differently. :-)

* Redesigned the Growatt class to manage various protocols. (v3.05 and v1.24 implemented)
* The new protocol definition allows user to define what modbus register should be read and which ones should be:
	* exported in JSON by calling `http://<ip>/status`
	* displayed on the UI `http://<ip>`
	* added to the graph in the UI
* MQTT_MAX_PACKET_SIZE doesn't need to be updated manually in the library. Instead the `PubSubClient::setBufferSize`has been used to do this dynamically.
* New `http://<ip>/uistatus` has been created to provide data for the UI
* UI changes:
	* Ui is generated dynamically based on the JSON provided.
	* Graph is now able to plot multiple values

Please test this before merging - I only have my inverter to test + don't have a mqtt set up yet!

zinserjan reviewed (commented):
Great work! I like the idea to support multiple protocols / more devices. 

I played around with it today, but I had to create my own procotol file + some other changes (see comments) for my setup:
- Growatt MIC 1000TL-X
- MQTT

<details>
 <summary>My version</summary>

**Growatt120.h**
```
#ifndef _GROWATT_1_20_H_
#define _GROWATT_1_20_H_

#include "Arduino.h"

#include "GrowattTypes.h"

// Growatt modbus protocol version 1.20
typedef enum {
    P120_INVERTER_STATUS, // Register 0
    P120_PV_POWER, // Register 1 + 2
    P120_PV1_VOLTAGE, // Register 3
    P120_PV1_CURRENT, // Register 4
    P120_PV1_POWER, // Register 5 + 6
    P120_AC_POWER, // Register 35 + 36
    P120_AC_VOLTAGE, // Register 38
    P120_AC_ENERGY_TODAY, // Register 53 + 54
    P120_AC_ENERGY_TOTAL, // Register 55 + 56
    P120_OPERATING_TIME, // Register 57 + 58
    P120_INVERTER_TEMPERATURE // Register 93
} eP120InputRegisters_t;

typedef enum {
    P120_OnOff, // Register 0
    P120_CMD_MEMORY_STATE, // Register 2
    P120_Active_P_Rate // Register 3
} eP120HoldingRegisters_t;

void init_growatt120(sProtocolDefinition_t &Protocol);

#endif // _GROWATT_1_20_H_
```

**Growatt120.cpp**
```
#include "Arduino.h"

#include "Growatt120.h"

void init_growatt120(sProtocolDefinition_t &Protocol) {
    // definition of input registers
    Protocol.InputRegisterCount = 11;

    // address, value, size, name, multiplier, unit, frontend, plot
    // Fragment 1: BEGIN
    Protocol.InputReadFragments[0] = sGrowattReadFragment_t{0, 6};
    Protocol.InputRegisters[P120_INVERTER_STATUS] = sGrowattModbusReg_t{0, 0, SIZE_16BIT, "InverterStatus", 1, NONE, true, false};
    Protocol.InputRegisters[P120_PV_POWER] = sGrowattModbusReg_t{1, 0, SIZE_32BIT, "PvPower", 0.1, POWER_W, true, true}; 
    Protocol.InputRegisters[P120_PV1_VOLTAGE] = sGrowattModbusReg_t{3, 0, SIZE_16BIT, "Pv1Voltage", 0.1, VOLTAGE, true, false};
    Protocol.InputRegisters[P120_PV1_CURRENT] = sGrowattModbusReg_t{4, 0, SIZE_16BIT, "Pv1Current", 0.1, CURRENT, true, false};
    Protocol.InputRegisters[P120_PV1_POWER] = sGrowattModbusReg_t{5, 0, SIZE_32BIT, "Pv1Power", 0.1, POWER_W, true, false};
    // Fragment 1: END
   
    // Fragment 2: BEGIN
    Protocol.InputReadFragments[1] = sGrowattReadFragment_t{30, 30};
    Protocol.InputRegisters[P120_AC_POWER] = sGrowattModbusReg_t{35, 0, SIZE_32BIT, "AcPower", 0.1, POWER_W, true, true}; 
    Protocol.InputRegisters[P120_AC_VOLTAGE] = sGrowattModbusReg_t{38, 0, SIZE_16BIT, "AcVoltage", 0.1, CURRENT, true, false}; 
    Protocol.InputRegisters[P120_AC_ENERGY_TODAY] = sGrowattModbusReg_t{53, 0, SIZE_32BIT, "AcEnergyToday", 0.1, CURRENT, true, false}; 
    Protocol.InputRegisters[P120_AC_ENERGY_TOTAL] = sGrowattModbusReg_t{55, 0, SIZE_32BIT, "AcEnergyTotal", 0.1, CURRENT, true, false}; 
    Protocol.InputRegisters[P120_OPERATING_TIME] = sGrowattModbusReg_t{57, 0, SIZE_32BIT, "OperatingTime", 0.5, CURRENT, true, false}; 
    // Fragment 2: END
 
    // Fragment 3: BEGIN
    Protocol.InputReadFragments[2] = sGrowattReadFragment_t{93, 1};
    Protocol.InputRegisters[P120_INVERTER_TEMPERATURE] = sGrowattModbusReg_t{93, 0, SIZE_16BIT, "InverterTemperature", 0.1, NONE, true, false}; 
    // Fragment 3: END
    Protocol.InputFragmentCount = 3;
    
    Protocol.HoldingRegisterCount = 3;
    Protocol.HoldingReadFragments[0] = sGrowattReadFragment_t{0, 64};
    Protocol.HoldingRegisters[P120_OnOff] = sGrowattModbusReg_t{0, 0, SIZE_16BIT, "OnOff", 1, NONE, true, false};
    Protocol.HoldingRegisters[P120_CMD_MEMORY_STATE] = sGrowattModbusReg_t{2, 0, SIZE_16BIT, "CmcMemoryState", 1, NONE, true, false};
    Protocol.HoldingRegisters[P120_Active_P_Rate] = sGrowattModbusReg_t{3, 0, SIZE_16BIT, "ActiveRate", 1, PRECENTAGE, true, false}; 

    Protocol.HoldingFragmentCount = 1;
}

```
</details>

What's missing for me is some way to postprocess the results or to be more specific to control the json output. For example, in the current version of the project the inverter status is tranformed into a human readable string instead of returning the number. I think there are two ways to do this transformation:
1. provide an option in the procotol configuration to configure the value/text mapping
2. remove the generic json creation and move the json generation into each procotol

crasu commented on SRC/ShineWiFi-ModBus/Growatt.cpp:
```diff
@@ -1,288 +1,322 @@
 
-      return true;
-    }else{
+  // read Input Registers
```
IMHO this function is pretty long already. Could be split up into ReadInputRegisters and ReadHoldingRegisters.

crasu reviewed (commented):
Like it ;-)

otti commented on SRC/ShineWiFi-ModBus/ShineWiFi-ModBus.ino:
```diff
@@ -318,6 +311,8 @@ void setup()
     WiFi.mode(WIFI_STA); // explicitly set mode, esp defaults to STA+AP
 
     #if MQTT_SUPPORTED == 1
+        MqttClient.setBufferSize(MQTT_PACKET_SIZE);
```
Works here with original ShineWifi-S Stick

zinserjan replied:
Btw I use the ShineWifi-X stick. But instead of plugging it into the inverter, I just plugged the Stick into an USB-Port of my PC to configure it. That makes the configuration and first test much easier for me :grinning:

tommarek replied:
as I said i can't really test this :( I need to setup my mqtt broker and try it. But this seems to have been fixed by the platformio build anyways.

zinserjan replied:
I think you can even test it without a MQTT broker. Just enable MQTT and go to the WiFi configuration by scan. In my case it didn't load in the past, but I need to test this again.

crasu replied:
The issue I have is that it just does not send mqtt messages. I had the feeling that the esp8266 is running out of memory ...

otti commented:
Hi @tommarek,

thanks for your amazing contribution.
I had to update ArduinoJson to the latest version (5 to 6) and also MqttPubSubClient because in my old version setBufferSize() was not implemented. This function will be a huge benefit for the usabillity.

I uploaded the firmware and everything seems to work basically. Unfortunately the sun went down after a few minutes of exploring all the new features. I will try to continue during the next days.

One thing I noticed, is that the structure and the scaling of the JSON data has changed. 
![grafik](https://user-images.githubusercontent.com/3578456/187268703-ea39b07e-4874-4593-8f2e-9ccfff5a0e19.png)
We schould keep the compatibility to old versions. Adding new entries should not be a problem, but changing the structure or the scaling of the values should be avoided.

The next thing I have to learn, is how to work with other contributors and how to deal with such complex pull requests. ;-)

Thanks for your PR and all the testing and comments from @crasu and @zinserjan

tommarek commented on SRC/ShineWiFi-ModBus/Growatt124.cpp:
```diff
@@ -0,0 +1,85 @@
+    // address, value, size, name, multiplier, unit, frontend, plot
+    // FEAGMENT 1: BEGIN
+    Protocol.InputRegisters[P124_I_STATUS] = sGrowattModbusReg_t{0, 0, SIZE_16BIT, "InverterStatus", 1, NONE, true, false}; // #1
+    Protocol.InputRegisters[P124_INPUT_POWER] = sGrowattModbusReg_t{1, 0, SIZE_32BIT, "Input power", 0.1, POWER_W, true, true}; // #2
```
I used the names from the growatt modbus protocol document. But you're right spaces are less than ideal. :-)

tommarek commented:
Hey guys, thanks for the comments. I'm on holiday now without access to my computer so I'll fix/answer everything on Monday. :-)

tommarek commented:
Re otti's comment: https://github.com/otti/Growatt_ShineWiFi-S/pull/30#issuecomment-1230701952

glad you like the change. I've implemented most of the requests from this feed into the PR and merged master again (uff :-) )

- The Json is now flattened so should be the same as before. Now the user needs to make sure to have different names in holding/input register definitons to avoid overwriting the values in the CSV output. But I reckon that's not a too big of an issue.
- the values were not scaled because of the bug in the protocol definition - the multipliers were all 1 :palmface:

In this PR i have removed the `AccumuatedEnergy` field alltogether as I didn't understand what it does and couldn't think of a way how to replicate that for my protocol. Could you please elaborate a bit on that - how it works and what it represents exactly please?

In the future I would like to change the protocol definitions to a class that could implement protocol specific data transformations - My `InverterStatus` is `5` when operating normally so I can't really use the old mapping you had before because it wouldn't work for all protocols. :(

Thanks
Tom

otti reviewed (approved):
Hi Tom,

I can compile this PR and upload it to my hardware. (Version 3.05) Everything works fine.
The JSON string can be processed by my backend without problems. So it is compatible to the previous version.
I haven't tested the version 1.24. But I think you have, because it is the new default ;-)
Things to do in future releases:
* The MQTT "last will" will still send "Status: "Disconnected" but it should send "InverterStatus: ???"
* The values in the web interface shall be rounded to a reasonable length.

Thanks again for your contribution.

tommarek commented:
I can add both of these changes in my branch, that's not a problem. Also what about the AccumulatedEnergy?

otti commented on SRC/ShineWiFi-ModBus/ShineWiFi-ModBus.ino:
```diff
@@ -148,7 +148,7 @@ String mqtttopic = "";
 String mqttpwd = "";
 
-char JsonString[MQTT_PACKET_SIZE] = "{\"Status\": \"Disconnected\" }";
+char JsonString[MQTT_PACKET_SIZE] = "{\"Status\": -1 }";
```
You renamed the "Status" in the MQTT string to "InverterStatus". I think we schould do the same over here.

otti commented:
Hi Tom,
`AccumulatedEnergy `: This feature was added by a colleague. At least the Growatt 1000S will store the total energy in steps of 0,1 kWh. During winter there are days where your production is below 0,1 kWh. So the inverter will not store the production of a whole day or even weeks. Every morning after start up the total energy falls back to a tenth of a kWh. (floor) If you draw a graph of the total energy, you will end up in a negative production every morning :-(
My colleague added this feature to have a counter with much better resolution. The downside is, that you have to update this register every morning after start up.
I asked him, if we have to keep this feature in the firmware. He is still using it. So it would be nice to keep it. But it it also OK for him if we remove it. He told me, that the firmware is running since multiple years without problems and he will "never" update it. :-)
So it is up to you, how we will go on with this feature.

otti commented:
Hi @tommarek ,
the UI tells me: `This branch cannot be rebased due to conflicts` .
I only have a basic knowledge about git/github. What do we have to do now?
If possible, I want to solve this in the UI. Pulling your branch and merging it to my local master, and than pushing directly to the origin/master seems wrong to me. Furthermore the origin/master should be protected againt direct pushing.

tommarek commented:
fixing it and will let you know when it's done :)

zinserjan commented:
@tommarek I tried to test this branch yesterday. Unfortunately I have still the same problem with the blank configuration page when MQTT is enabled. Without MQTT it seems to work. Master is working, too. I would like to find the cause for this, but I don't have any experience in debugging this ESP stuff... In my case it's the ShineWifi-X. @crasu does it work for you? If I'm correct you had the same problem the last time.

Regarding my comment about the missing protocol, that's probably v1.05. But I think I can add this later, as it seems that this version is used on my Growatt MIC 1000TL-X. 

One questions came to my mind yesterday, is is possible to detect the protocol version automatically?

tommarek commented:
@otti  I tried to do the rebase but man it's complicated I think the easiest way would be to merge my branch manually and I will make sure my future branches are possible to rebase from the UI, sorry for this. The branch has all conflict resolved by me when merging upstream/master into my branch and all the conflicts are resolved. Please don't merge it just yet until i test the MQTT issue and fix it.
@zinserjan Hey, I will setup my MQTT today and will test this out this evening will keep you posted. Will try to resolve this asap

zinserjan commented:
@tommarek rebasing is kind of hard due to amount of commits. I would recommend to sync the master branch in your fork with this repository from @otti. After that you can merge your remote master into this branch and resolve the merge conflics once. If you need any help, I think I can do it after the work in the evening and send you a PR.

zinserjan commented:
@tommarek that's good to hear that you have the same problem. 

Btw I tried to find the cause this evening. I flashed my stick around 50 times. The issue happens sometimes for me even without MQTT, but less often.

I removed nearly everything until it starts to work again... After I removed the following code the WiFi scan page seems to work:

```
// Read Inverter every REFRESH_TIMER ms [defined in config.h]
    // ------------------------------------------------------------
    if ((now - RefreshTimer) > REFRESH_TIMER)
    {
....
    }
```

Adding this leads to the error, removing it fixed it. Justed tested it again. Maybe this happens only, because I use the stick without the inverter to do the initial configuration. Anyway, I gave up for today :D

tommarek commented:
finally found the issue (i hope lol) - buffer size 4096 caused it. I shrinked it down to 2048 which should still be plenty/ I will add rounding of the values just to save some space in the output buffer and remove the rounding from the UI tomorrow. please check if this commit fixed it for you, thanks!

zinserjan commented:
@tommarek I just tried it with your settings. Still didn't work for me, but I tested it with the changes of my PR tommarek/Growatt_ShineWiFi-S#1 and this seems to be the problem. I think there is not enough memory for all of this. As soon as I remove the inclusion of the other procotols the WiFi page works.

tommarek commented:
@zinserjan please try the latest version, I've changed the logic to only compile the protocol files that are actually used (selected in config.h)

zinserjan commented:
@tommarek awesome, it's working!
- without inverter -> working
- with inverter -> working
- with mqtt -> working
- switching back to config portal -> working

zinserjan commented on README.md:
```diff
@@ -152,3 +159,16 @@ If the total energy is 0.199 kWh before sunset, the totoal enrgy will be reset t
+	* exported in JSON by calling `http://<ip>/status`
+	* displayed on the UI `http://<ip>`
+	* added to the graph in the UI
+* MQTT_MAX_PACKET_SIZE doesn't need to be updated manually in the library. Instead the `PubSubClient::setBufferSize`has been used to do this dynamically.
```
After we found the memory issues. Maybe we can add setBufferSize back.

zinserjan commented on README.md:
```diff
@@ -60,17 +61,20 @@ For IoT applications the raw data can now read in JSON format (application/json)
+        "AccumulatedEnergy": 320,
+        "Cnt": 333
+      },
+      "holding": {}
```
I think this example is incorrect.

zinserjan commented on SRC/ShineWiFi-ModBus/ShineWiFi-ModBus.ino:
```diff
@@ -145,7 +143,7 @@ String mqtttopic = "";
 String mqttpwd = "";
 
-char JsonString[MQTT_MAX_PACKET_SIZE] = "{\"Status\": \"Disconnected\" }";
+char JsonString[MQTT_MAX_PACKET_SIZE] = "{\"InverterStatus\": -1 }";
```
Maybe this fallback string should be defined within the different protocols, because the name may be different?
The same applies for the mqtt last will message.

crasu reviewed (approved):
Just two minor comments regarding documentation. Tested this on my esp32. Looks good. We should get this merged as this pr is far to big already.

tommarek commented:
Thanks guys! I'm not sure how to make the rebase work so if anyone has any idea I'd apreciate the help. 

Things to do after this gets merged (suggestions):
- Chose a coding style (variable naming, brackets on new lines vs. at the end of lines [I prefer that :D], etc.)  and update the entire project before it gets bigger
- Change protocol definition so it can
  - translate protocol specific constants to strings (ie. InverterStatus 1 => "normal")
  - control the inverter - interface, operations, etc.
    - Idea here is that with current energy prices it might be profitable to buy energy when it's cheap (store it in the battery) and then optionally sell it when the price is at the peak. (checked the prices and the difference in price is [massive](https://www.ote-cr.cz/en/short-term-markets/electricity/day-ahead-market?set_language=en&date=2022-02-28)
- Nicer UI

zinserjan commented:
> I'm not sure how to make the rebase work so if anyone has any idea I'd apreciate the help.

`This branch has no conflicts with the base branch` --> Nothing to do for you.

As far as I remember @otti has three options to merge this:
- fast forward or rebase -> put all commits on top of the master (highest conflict potential)  
- via merge commit -> include all commits into the master (should already work cause there a no conflicts at the moment)
- squash merge -> include the changes with a single commit (my recommendation, cause this simplifies the history)

zinserjan commented on SRC/ShineWiFi-ModBus/Config.h.example:
```diff
@@ -3,7 +3,7 @@
 
 // Define used modbus version used by your invertor here. Currently
-// only 124 (v1.24) and 305 (v3.05) versions are supported.
+// only 120 (v1.24), 124 (v1.24) and 305 (v3.05) versions are supported.
```
should be 120 (v1.20)

otti commented:
Hi @crasu  and @tommarek,

I just did the squash mege on this PR.

´Chose a coding style...` perhaps we can just add a .`clang-format` to the project. I personally like the brackets on new lines. But I don't  insist on them :-) Feel free to add a coding schema file to the project. Perhaps there is one schema out there used by other big open source projects.

Thanks again for all your work and testing.

tommarek commented:
@zinserjan already had a schema added to his PR to generic_protocol branch so he might want to do this (?) Regarding the brackets I don't mind either way I just like neatly formatted code. :-) in my next pr i will focus on updating the protocol so it offers a bit more flexibility. Need to think about the architecture though... Wish C++ had metaclasses. :-)

zinserjan commented:
@tommarek OK, I just created an PR with my wip config.
